segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/filesystem/FileSystemStorage.java

```diff
@@ -0,0 +1,407 @@
+ * writing the same data at the same offset till the time the earlier owner gets a notification that it is not the
+ * current owner. Once the earlier owner received this notification, it stops writing to the segment.
+ *
+ */
+@Slf4j
+public class FileSystemStorage implements Storage {
+
+    //region members
+
+    private final FileSystemStorageConfig config;
+    private final ExecutorService executor;
+    private final AtomicBoolean closed;
+
+    //endregion
+
+    //region constructor
+
+    /**
+     * Creates a new instance of the FileSystemStorage class.
+     *
+     * @param config   The configuration to use.
+     * @param executor The executor to use for running async operations.
+     */
+    public FileSystemStorage(FileSystemStorageConfig config, ExecutorService executor) {
+        Preconditions.checkNotNull(config, "config");
+        Preconditions.checkNotNull(executor, "executor");
+        this.closed = new AtomicBoolean(false);
+        this.config = config;
+        this.executor = executor;
+    }
+
+    //endregion
+
+    //region Storage implementation
+
+    /**
+     * Initialize is a no op here as we do not need a locking mechanism in case of file system write.
+     * @param containerEpoch The Container Epoch to initialize with (ignored here).
+     */
+    @Override
+    public void initialize(long containerEpoch) {
+        this.ensureInitializedAndNotClosed();
+    }
+
+    @Override
+    public CompletableFuture<SegmentHandle> openRead(String streamSegmentName) {
+        return supplyAsync( streamSegmentName, () -> syncOpenRead(streamSegmentName));
+    }
+
+
+    @Override
+    public CompletableFuture<Integer> read(SegmentHandle handle, long offset, byte[] buffer, int bufferOffset, int
+            length, Duration timeout) {
+        return supplyAsync( handle.getSegmentName(), () -> syncRead(handle, offset, buffer, bufferOffset, length));
+    }
+
+    @Override
+    public CompletableFuture<SegmentProperties> getStreamSegmentInfo(String streamSegmentName, Duration timeout) {
+        return supplyAsync(streamSegmentName, () -> syncGetStreamSegmentInfo(streamSegmentName));
+    }
+
+    @Override
+    public CompletableFuture<Boolean> exists(String streamSegmentName, Duration timeout) {
+        return supplyAsync(streamSegmentName, () -> syncExists(streamSegmentName));
+    }
+
+    @Override
+    public CompletableFuture<SegmentHandle> openWrite(String streamSegmentName) {
+        return supplyAsync(streamSegmentName, () -> syncOpenWrite(streamSegmentName));
+    }
+
+    @Override
+    public CompletableFuture<SegmentProperties> create(String streamSegmentName, Duration timeout) {
+        return supplyAsync( streamSegmentName, () -> syncCreate(streamSegmentName));
+    }
+
+    @Override
+    public CompletableFuture<Void> write(SegmentHandle handle, long offset, InputStream data, int length, Duration
```

Comment:
Given that we only append to files, I suppose the writes will be issued in append order, but the execution might not follow that order because the writes are assigned to a thread in the executor pool. This is likely to create gaps that will be filled over time, so I'm wondering two things:

- Is there any risk of a race where we hit one such a gap while reading?
- Should we consider pre-allocating space for files to avoid hiccups due to frequent file metadata updates? This is a pretty common technique to speed up sequential writes.